### PR TITLE
ARROW-4804: [Rust] Parse Date32 and Date64 in CSV reader

### PR DESCRIPTION
**Pull request by Dandandan**
This is based on https://github.com/apache/arrow/pull/8611 by @Jibbow and some suggestions by @alamb 

Adds date32 / date64 to the csv reader. This also fixes the benchmark which now includes date types which were added by @seddonm1

There are some missing parts in the date format support (such as actual ms support) but those can be implemented I think as separate PRs.

### CODE REVIEW

File: rust/arrow/test/data/various_types.csv [hunk 1]
```diff
@@ -1,6 +1,6 @@
-c_int|c_float|c_string|c_bool
-1|1.1|"1.11"|true
-2|2.2|"2.22"|true
-3||"3.33"|true
-4|4.4||false
-5|6.6|""|false
\ No newline at end of file
+c_int|c_float|c_string|c_bool|c_date
```
Comment (alamb) [inline rust/arrow/test/data/various_types.csv:1]: I suggest we also add a c_datetime column too to cover the schema inference of that type too

**Review by alamb — APPROVED** [review]
Looks good to me. Thank you @Dandandan

**Comment by seddonm1** [comment]
This does highlight a mistake made by me when implementing the `CAST` to Date64. I will do a PR to address and align.

**Comment by Dandandan** [comment]
Pushed a speed improvement by avoiding parsing of the format (and just using the default formatting).

File: rust/arrow/test/data/various_types.csv (hunk 1 above)
Comment (Dandandan) [reply rust/arrow/test/data/various_types.csv]: Added, added some extra tests / checks as well.

**Comment by seddonm1** [comment]
This is really nice @Dandandan . I will have to do some reading on the implementation of the parse trait by the chrono library.

**Review by alamb — APPROVED** [review]
Looks good -- thanks @Dandandan

**Pull request closed without merging**